### DISCUSSION TITLE
Using CLI with compiled JavaScript

### DISCUSSION THREAD

**Discussion by Hannes-Kunnen** [mikro-orm/mikro-orm#3152] [Q&A] [answered]
**Describe the bug**
I'm not sure if this is a bug or I'm just misinterpreting the documentation. I'm trying to run migrations in production with only the JavaScript version of the migration files (I have both paths set in the settings). When I try to run the command `mikro-orm migration:up` I receive an error that no entities are discovered (see stack trace). However when I run the `mikro-orm debug` command it states that the entities files are found and that these will be used.

In the metadata providers part of the documentation I read the following:
> If we use folder-based discovery, we should specify paths to the compiled entities via `entities` as well as paths to the TS source files of those entities via `entitiesTs`. When we run the ORM via `ts-node`, the latter will be used automatically, or if we explicitly pass `tsNode: true` in the config. Note that `tsNode: true` should not be part of production config.

I don't have the `tsNode`variable set and both `entities` and `entitiesTs` are specified, so I would assume this isn't a problem.

I've searched some issues and the only ones I think are truly related are https://github.com/mikro-orm/mikro-orm/issues/1339 and https://github.com/mikro-orm/mikro-orm/issues/1957 but I can't find a definitive answer.

I currently use the workaround of setting the `ENV MIKRO_ORM_CLI_USE_TS_NODE` environment variable to `false`. When I do this everything seems to be working fine,.

**Stack trace**
```
ts-node not installed, support for working with TS files might not work
mikro-orm migration:up

Migrate up to the latest version

Options:
  -t, --to       Migrate up to specific version                         [string]
  -f, --from     Start migration from specific version                  [string]
  -o, --only     Migrate only specified versions                        [string]
  -v, --version  Show version number                                   [boolean]
  -h, --help     Show help                                             [boolean]

MetadataError: No entities were discovered
    at Function.noEntityDiscovered (/project/node_modules/@mikro-orm/core/errors.js:151:16)
    at MetadataValidator.validateDiscovered (/project/node_modules/@mikro-orm/core/metadata/MetadataValidator.js:38:42)
    at MetadataDiscovery.findEntities (/project/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:91:24)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async MetadataDiscovery.discover (/project/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:34:9)
    at async MikroORM.discoverEntities (/project/node_modules/@mikro-orm/core/MikroORM.js:90:25)
    at async Function.init (/project/node_modules/@mikro-orm/core/MikroORM.js:45:9)
    at async Function.handleMigrationCommand (/project/node_modules/@mikro-orm/cli/commands/MigrationCommandFactory.js:92:21) {
  entity: undefined
}
error Command failed with exit code 1.
```
```
ts-node not installed, support for working with TS files might not work
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 5.1.4
   - node 16.13.0
   - knex 2.0.0
   - pg 8.7.3
   - typescript not-found
 - package.json found
 - ts-node enabled
 - searched config paths:
   - /project/dist/src/mikro-orm.config.js (found)
   - /project/mikro-orm.config.js (not found)
 - configuration found
 - will use `entities` array (contains 0 references and 1 paths)
   - /project/dist/src/entities/**/*.model.js (found)
 - could use `entitiesTs` array (contains 0 references and 1 paths)
   - /project/dist/src/entities/**/*.model.ts (not found)
```

**To Reproduce**
1. Create project with TypeScript entities and migrations
2. Compile project
3. Run `mikro-orm migration:up` without the typescript files and ts-node present

**Expected behavior**
Since ts-node is not installed and only JavaScript entities are discovered these files should be used and the migration should work fine.

**Additional context**
I run these commands in containers, but I don't think that is related as I can just reproduces these errors in a terminal.

**Versions**

| Dependency | Version |
| - | - |
| node | 16.13.0 |
| typescript | 4.4.4 |
| mikro-orm | 5.1.4 |
| your-driver | PostgreSQL |

**Comment by B4nan** [comment, marked as answer]
The CLI operates based on the `useTsNode` flag in package.json, so if you have that enabled, it will work in the TS mode - you can see it tries to register it in the warning on first line. I guess we should ignore that flag automatically if we see ts-node is not installed (we fail to register it). 

You can dynamically override the value via `MIKRO_ORM_CLI_USE_TS_NODE` env var too, but that's pretty much the same as what you are doing now.

**Reply by Hannes-Kunnen** [reply]
Thanks for the clarification, I think it would be nice indeed to ignore that flag so there is no need to override it using an environment variable.

For now I'll just continue using it like I did before.

**Reply by B4nan** [reply]
This should do it: https://github.com/mikro-orm/mikro-orm/commit/457d9d32d78ed34b254aae3b2b2f29b382e7175e

**Reply by Hannes-Kunnen** [reply]
Any idea when this will be released?

**Reply by B4nan** [reply]
You can use the dev version, every commit to master is released.
